core/src/main/java/org/pentaho/di/connections/ConnectionManager.java

```diff
@@ -102,33 +130,41 @@ public synchronized void reset() {
   /**
    * This getter should not generally be used because it is limited to the global scope. It would be better to have
    * almost all callers use Bowl.getConnectionManager().
-   *
+   * <p>
    * This instance may still be used to register ConnectionProviders and Lookup Filters.
    *
    * @return ConnectionManager
    */
+  @NonNull
   public static ConnectionManager getInstance() {
+    if ( instance == null ) {
+      instance = new ConnectionManager();
+    }
+
```

Comment:
I think this is now not thread safe. You should be able to go back to the old version, where the instance is assigned to the variable up on line 56. 

I'm kindof surprised this compiles... you aren't supposed to be allowed to assign final variables in an unsafe manner. 

OH! you made it not final! interesting. 

so A) why?  and B) definitely not thread safe.

Reply:
Trying to avoid a possible static initialization cycle between the two classes `DefaultBowl` and `ConnectionManager`. 

`ConnectionManager` had statically initialized field that requires `DefaultBowl` to have been statically initialized.
While `DefaultBowl` has methods of type `ConnectionManager` which could cause it to have to initialize as well.

This may not be an issue.

Regarding thread-safety, I was admittedly lazy thinking to myself it wouldn't be a big issue...

Reply:
I've definitely seen race conditions in `ConnectionManager.getInstance()` specifically. 

This is partly why DefaultBowl uses lazy initialization for the manager and metastore.